waspc/src/Wasp/Generator/WriteFileDrafts.hs

```diff
@@ -60,6 +64,12 @@ type RelPathsToChecksums = [(FileOrDirPathRelativeTo ProjectRootDir, Checksum)]
 
 type RelPathsToChecksumsMap = Map.HashMap (FileOrDirPathRelativeTo ProjectRootDir) Checksum
 
+-- | Takes file drafts and verifies if the destination paths are unique or not
+checkIfUniqueDestPaths :: [FileDraft] -> IO ()
+checkIfUniqueDestPaths fileDrafts = do
+  let fileDestPaths = map (\fd -> (fd,) <$> getDstPath fd) fileDrafts
+  if length (nub fileDestPaths) == length fileDestPaths then (putStrLn "") else (throwString "Destination paths are not unique.")
```

Comment:
This is a really nice job! I was thinking it would be cool if we could also show the actual duplicated `FileDraft`s, but when I wanted to do something like either of the following, I was quickly reminded just how limited our `FileDraft` type really is :D 

```haskell
-- No Hashable
duplicateFdDestPaths = fdDestPaths \\ (Set.toList . Set.fromList $ fdDestPaths)

-- No Ord
duplicateFdDestPaths = concat $  filter ((> 1) . length) (group . sort $ fdDestPaths)
```

So I think just throwing an error without naming the offender(s) sounds great for now!

Reply:
A couple minor edits to consider would be using `error` (since we typically just do that in our code so it may make it easier to grep for them all moving forward) as well as using `unless`:

```haskell
-- | Takes file drafts and verifies if the destination paths are unique or not.
-- If there are duplicates, we throw an error since this is a Wasp bug that needs to be fixed.
checkIfUniqueDestPaths:: [FileDraft] -> IO ()
checkIfUniqueDestPaths fileDrafts =
  let fileDestPaths = map (\fd -> (fd,) <$> getDstPath fd) fileDrafts
   in unless (length (nub fileDestPaths) == length fileDestPaths) (error "FileDraft destination paths are not unique!")
```

Also, in instances like this I think we often would say "ensure", so maybe calling the function something like `ensureDestPathsAreUnique` perhaps? 🤔 Thoughts?

Reply:
No sweat! Two other notes:

1) You can nix the `do` since we are using a single expression.

2) One thing I missed yesterday - I don't think you need the tuple section part `(fd,) <$>` (unless there was a reason for this?). That ends up giving you a type like:
```haskell
fileDestPaths :: [Either
   (Path' (Rel ProjectRootDir) File')
   (FileDraft, Path' (Rel ProjectRootDir) Dir')]
```

I think we can get away with this:
```haskell
ensureDestPathsAreUnique :: [FileDraft] -> IO ()
ensureDestPathsAreUnique fileDrafts =
  let fileDestPaths = map getDstPath fileDrafts
   in unless (length (nub fileDestPaths) == length fileDestPaths) (error "FileDraft destination paths are not unique!")
```

Please let me know if you have any questions on the test part. And nice job, keep it up!

Reply:
Looking good!
1. Error -> I think it is right to call `error` here, since it is fatal error (bug in wasp compiler) that we don't want to recover from and want to crash immediately. `error` is the right choice for that! `throwError` would be more interesting if we wanted to possibly recover in some way, due to user-caused error.
2. Does `ensureDestPathsAreUnique` have to be `IO ()`? Looking at it, I would say it can be just `()`.
3. @shayneczyzewski I agree it would be cool if error actually gave us a bit more info (although this is also great already)! We don't have Hashable or Ord on FileDraft, but anyway we are not dealing with filedrafts, we are dealing with destination paths which are of type StrongPath, right? And StrongPath, if I am correct, has both Ord and Hashable defined on it. So we should be able to perform a bit fancier stuff on it!

Reply:
Naming-wise -> I usually use `ensure` if function checks that some conditions is satisfied, and in case when it is not satisfied, it tries to satisfy it by possibly doing some mutation / effect -> e.g. ensureDatabaseIsConnected where it connects to database if it is not connected yet. I usually use `check` when function just checks, without trying to do anything if condition is not met -> e.g. checkIfDatabaseIsConnected, which just returns True/False and doesn't try to do anything.
I think this function is right in between, because it doesn't just return True/False, it actually does something to ensure files are unique, which is throw error :D. But it is a bit weird because it doesn't really try to fix -> well, the fix is to throw an error and have developer fix it :D.

What about another idea: `assert`? [assert](https://en.wikipedia.org/wiki/Assertion_(software_development)) is normally used to check a condition and crash / throw error if conditions is not met.
So while both `ensure` and `check` are good enough I think, I would probably go with `assert`.

Reply:
Thanks for the feedback @Martinsos! Replies:

1) Agreed, no handling, so `error` feels right.
2) I thought the same after writing, but since we simplified the code with `unless` we needed it to be Applicative at least. We could make it something like: `ensureDestPathsAreUnique :: Applicative f => [FileDraft] -> f ()`. Or we could remove the use of `unless` to be `()`, but the downside is we then have to _strictly force_ the invocation to ensure it actually gets called (i.e., by enabling bang patterns and doing `let !_ = assertDestPathsAreUnique fileDrafts`, or some other way?).
3) You are right! 🤦🏻‍♂️ arg When I did that testing, it was when there was still a use of `(fd,) <$>` that I noted above (which included `FileDraft` in the type). But these are indeed just `StrongPath` types. So @InfiniteVerma now that the `map` was updated, we can do some of the following (assuming we go assert):

```haskell
let !_ = assertDestPathsAreUnique fileDrafts

-- | Takes file drafts and verifies if the destination paths are unique.
assertDestPathsAreUnique :: [FileDraft] -> ()
assertDestPathsAreUnique fileDrafts =
  let fdDestPaths = map getDstPath fileDrafts
      duplicateFdDestPaths = map head $ filter ((> 1) . length) (group . sort $ fdDestPaths)
      errMessage = unlines $ "FileDraft destination paths are not unique! Duplicates include: " : map show duplicateFdDestPaths
   in if null duplicateFdDestPaths then () else error errMessage
```

Thoughts @Martinsos? And remember, this is just an option I'm providing to help jumpstart you @InfiniteVerma so feel free to tweak this and play and come back to us with what you think makes sense (i.e. using `unless` or not, etc). Can't wait to see what you decide on and the test! Thanks for sticking with this!

Reply:
Code is looking good to me!

I haven't considered laziness I have to admit, that is a very good observation. You think it won't execute it all, due to us not using the returned value, and function having no side effects (IO)? Uff!
OK, I did quick test in ghci:
I defined
```hs
f x = if x == True then () else error "I expected True"
```
and then I defined
```hs
g x = do; return (f x); putStrLn "wohooo"
```

`g False` then gave me `wohooo`, which means it never evaluated (f x) -> yikes!

But then I changed it to
```hs
g x = do; return $! f x; putStrLn "wohooo"
```
`g False` now crashed with an error, which is what we wanted.

I also tried with `let !_` and that also works correctly btw, I am not sure why I went with return, just had it first on my mind. `return` approach doesn't require an extension though.

Reply:
Ah yes, good point, you could also use `return $! assertDestPathsAreUnique fileDrafts` to call it instead 👍🏻

Reply:
Hey, @shayneczyzewski sorry for the late reply! I've refactored the code with the comments above. 

Putting it as ```assert``` does make more sense @Martinsos, thanks for clearing that up! 

Also, having it ```()``` and not ```IO ()``` is better. It's interesting how we need to strictly force it now. Thanks for explaining that part as well!

Reply:
Awesome, thanks @InfiniteVerma. Apologies for my delay, as we had a team retreat last week. I will do the final review this week though, thanks! 🙏🏻